cartographer_grpc/framework/rpc.cc

```diff
@@ -269,10 +220,56 @@ bool* Rpc::GetRpcEventState(Event event) {
       return &read_event_pending_;
     case Event::WRITE:
       return &write_event_pending_;
+    case Event::WRITE_NEEDED:
+      return &write_needed_event_pending_;
   }
   LOG(FATAL) << "Never reached.";
 }
 
+void Rpc::EnqueueMessage(SendItem&& send_item) {
+  cartographer::common::MutexLocker locker(&send_queue_lock_);
+  send_queue_.emplace(SendItem{std::move(send_item.msg), send_item.status});
+}
+
+void Rpc::PerformFinish(std::unique_ptr<::google::protobuf::Message> message,
+                        ::grpc::Status status) {
+  SetRpcEventState(Event::FINISH, true);
+  switch (rpc_handler_info_.rpc_type) {
+    case ::grpc::internal::RpcMethod::BIDI_STREAMING:
+      CHECK(!message);
+      server_async_reader_writer_->Finish(
+          status,
+          new RpcEvent{Event::FINISH, weak_ptr_factory_(this), event_queue_});
+      break;
+    case ::grpc::internal::RpcMethod::CLIENT_STREAMING:
+      response_ = std::move(message);
+      SendUnaryFinish(
+          server_async_reader_.get(), status, response_.get(),
+          new RpcEvent{Event::FINISH, weak_ptr_factory_(this), event_queue_});
+      break;
+    case ::grpc::internal::RpcMethod::NORMAL_RPC:
+      response_ = std::move(message);
+      SendUnaryFinish(
+          server_async_response_writer_.get(), status, response_.get(),
+          new RpcEvent{Event::FINISH, weak_ptr_factory_(this), event_queue_});
+      break;
+    default:
+      LOG(FATAL) << "RPC type not implemented.";
+  }
+}
+
+void Rpc::PerformWrite(std::unique_ptr<::google::protobuf::Message> message,
+                       ::grpc::Status status) {
+  CHECK_NE(rpc_handler_info_.rpc_type, ::grpc::internal::RpcMethod::NORMAL_RPC);
+  CHECK_NE(rpc_handler_info_.rpc_type,
+           ::grpc::internal::RpcMethod::CLIENT_STREAMING);
+  SetRpcEventState(Event::WRITE, true);
+  response_ = std::move(message);
```

Comment:
Before the refactor this could also be a nullptr, now no longer?

Reply:
It still can `std::move(nullptr)` clears the thing on the LHS.

Reply:
Sorry I should have read your comment more carefully: `PerformWrite` must be called with a non-null message, which is ensured in `PerformWriteIfNeeded`. I have added a `CHECK` for that.